scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -118,261 +79,409 @@ class EscapeHatch(
-    def trackUnusedRules(
-        rules: String,
-        toogle: Toogle,
-        anchorPosition: AnchorPosition): Unit = {
-      toogle match {
-        case Toogle.Disable =>
-          currentlyDisabledRules = currentlyDisabledRules ++ splitRules(rules)
-
-        case Toogle.Enable =>
-          val enabledRules = splitRules(rules)
-          val enabledNotDisabled = enabledRules -- currentlyDisabledRules
-
-          enabledNotDisabled.foreach(
-            rule => unusedEnable += unusedEnableWarning.at(anchorPosition.value)
-          )
+      tree.foreach {
+        case t @ Defn.Class(mods, _, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Defn.Object(mods, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Defn.Trait(mods, _, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Defn.Type(mods, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Defn.Def(mods, _, _, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Defn.Val(mods, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Defn.Var(mods, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Term.Param(mods, _, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Ctor.Primary(mods, _, _) if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case t @ Ctor.Secondary(mods, _, _, _, _)
+            if hasSuppressWarnings(mods) =>
+          addAnnotatedEscape(t, mods)
+
+        case _ => ()
       }
+
+      new AnnotatedEscapes(builder.result())
     }
 
-    tree.foreach { t =>
-      associatedComments.trailing(t).foreach {
-        // matches simple expressions
-        //
-        // val a = (
-        //   1,
-        //   2
-        // ) // scalafix:ok RuleA
-        //
-        case comment @ Token.Comment(FilterExpression(rules)) =>
-          val anchorPosition = AnchorPosition(comment.pos)
-          if (!visitedFilterExpression.contains(anchorPosition)) {
-            escapes += Escape.Expression(
-              ExpressionPosition(t.pos),
-              anchorPosition,
-              rules
-            )
-            visitedFilterExpression += anchorPosition
-          }
+    private def hasSuppressWarnings(mods: List[Mod]): Boolean =
+      mods.exists {
+        case Mod.Annot(Init(Type.Name(SuppressWarnings), _, _)) => true
+        case Mod.Annot(
+            Init(Type.Select(_, Type.Name(SuppressWarnings)), _, _)) =>
+          true
+        case _ => false
+      }
 
-        case _ => ()
+    private def extractRules(mods: List[Mod]): List[(String, Position)] = {
+      def process(rules: List[Term]) = rules.map {
+        case lit @ Lit.String(rule) =>
```

Comment:
What happens if this is not a string literal?

```
val x = "foo"
@SuppressWarnings(Array(x))
```

I think it's fair to require string literals, but we should at least not crash if there is a variable.

Reply:
oops. That's another corner case that I did not consider. I think there is no a way to lookup the value. Am I right? Just avoiding crashing for now unless you have some better idea.